pemFioi/quickpi/blocklyQuickPi_lib.js

```diff
@@ -8826,10 +8832,65 @@ var getContext = function (display, infos, curLevel) {
         context.runner.waitDelay(callback, millis);
     };
 
-    context.quickpi.getTemperature = function(location, callback) {
-        var retVal =  25;
 
-        context.waitDelay(callback, retVal);
+    // TODO: change url to the "prod" one
+    context.quickpi.getTemperatureFromCloudUrl = "https://mapadev.com/nicolastest/weather.php";
+
+    // setup the supported towns
+    $.get(context.quickpi.getTemperatureFromCloudUrl + "?q=" + "supportedtowns", function(towns) {
+        context.quickpi.getTemperatureFromCloudSupportedTowns = towns;
+    });
+
+    // We create a cache so there is less calls to the api and we get the results of the temperature faster
+    context.quickpi.getTemperatureFromCloudCache = [];
+
+    context.quickpi.getTemperatureFromCloud = function(location, callback) {
+
+        function _updateGetTemperatureFromCloudCache(cache, location, newVal) {
+            for (var i = 0; i < cache.length; i++) {
+                var curr = cache[i];
+                if (curr.location === location) {
+                    curr.lastUpdate = Date.now();
+                    curr.temperature = newVal;
+                    return;
+                }
+            }
+            cache.push({
+                location: location,
+                lastUpdate: Date.now(),
+                temperature: newVal
+            });
+        }
+
+        var url = context.quickpi.getTemperatureFromCloudUrl;
+
+        if (!context.quickpi.getTemperatureFromCloudSupportedTowns.includes(location))
+            throw strings.messages.getTemperatureFromCloudWrongValue.format(location);
+
+        var cache = context.quickpi.getTemperatureFromCloudCache;
+        for (var i = 0; i < cache.length; i++) {
```

Comment:
Why not make the cache an object whose keys are the locations?

Reply:
Hello, I tried to do that at first but did not succeeded.

Now I fixed what didn't worked when I tried with an object and it is working well.